test/catch2_test_warp_store.cu

```diff
@@ -0,0 +1,313 @@
+          int ITEMS_PER_THREAD,
+          int TOTAL_WARPS,
+          typename T,
+          typename OutputIteratorT,
+          typename ActionT>
+__global__ void warp_store_kernel(OutputIteratorT output_iterator, ActionT action)
+{
+  using warp_store_t = cub::WarpStore<T, ITEMS_PER_THREAD, StoreAlgorithm, LOGICAL_WARP_THREADS>;
+  using storage_t    = typename warp_store_t::TempStorage;
+
+  constexpr int tile_size = ITEMS_PER_THREAD * LOGICAL_WARP_THREADS;
+  __shared__ storage_t storage[TOTAL_WARPS];
+
+  const int tid = cub::RowMajorTid(blockDim.x, blockDim.y, blockDim.z);
+  T reg[ITEMS_PER_THREAD];
+
+  for (int item = 0; item < ITEMS_PER_THREAD; item++)
+  {
+    reg[item] = static_cast<T>(tid * ITEMS_PER_THREAD + item);
+  }
+
+  const int warp_id = tid / LOGICAL_WARP_THREADS;
+  warp_store_t store(storage[warp_id]);
+
+  action(store, output_iterator + (warp_id * tile_size), reg);
+}
+
+template <cub::WarpStoreAlgorithm StoreAlgorithm,
+          int LOGICAL_WARP_THREADS,
+          int ITEMS_PER_THREAD,
+          int TOTAL_WARPS,
+          typename T,
+          typename OutputIteratorT,
+          typename ActionT>
+void warp_store(OutputIteratorT output_iterator, ActionT action)
+{
+  warp_store_kernel<StoreAlgorithm,
+                    LOGICAL_WARP_THREADS,
+                    ITEMS_PER_THREAD,
+                    TOTAL_WARPS,
+                    T,
+                    OutputIteratorT,
+                    ActionT><<<1, TOTAL_WARPS * LOGICAL_WARP_THREADS>>>(output_iterator, action);
+}
+
+struct guarded_store_t
+{
+  int valid_items;
+  template <cub::WarpStoreAlgorithm StoreAlgorithm,
+            int LOGICAL_WARP_THREADS,
+            int ITEMS_PER_THREAD,
+            typename T,
+            typename OutputIteratorT>
+  __device__ void
+  operator()(cub::WarpStore<T, ITEMS_PER_THREAD, StoreAlgorithm, LOGICAL_WARP_THREADS> store,
+             OutputIteratorT output,
+             T (&reg)[ITEMS_PER_THREAD])
+  {
+    store.Store(output, reg, valid_items);
+  }
+};
+
+struct unguarded_store_t
+{
+  template <cub::WarpStoreAlgorithm StoreAlgorithm,
+            int LOGICAL_WARP_THREADS,
+            int ITEMS_PER_THREAD,
+            typename T,
+            typename OutputIteratorT>
+  __device__ void
+  operator()(cub::WarpStore<T, ITEMS_PER_THREAD, StoreAlgorithm, LOGICAL_WARP_THREADS> store,
+             OutputIteratorT output,
+             T (&reg)[ITEMS_PER_THREAD])
+  {
+    store.Store(output, reg);
+  }
+};
+
+template <cub::WarpStoreAlgorithm StoreAlgorithm,
+          int LOGICAL_WARP_THREADS,
+          int ITEMS_PER_THREAD,
+          int TOTAL_WARPS,
+          typename T>
+thrust::device_vector<T> compute_reference(int valid_items)
+{
+  const int tile_size        = LOGICAL_WARP_THREADS * ITEMS_PER_THREAD;
+  const int total_item_count = TOTAL_WARPS * tile_size;
+  thrust::device_vector<T> d_input(total_item_count);
+
+  if (StoreAlgorithm == cub::WarpStoreAlgorithm::WARP_STORE_STRIPED)
```

Comment:
Some compilers emit warnings when the condition is compile-time known. Let's use `CUB_IF_CONSTEXPR` in such places.